// randomizer/lib/randomizer.js
var moment = require('moment');

Randomizer = function () {
    this.Math          = Math;
    this.moment        = moment;
    this.x             = null;
    this.y             = null;
    this.isExecute     = true;
    this.dayTimeStarts = 8;
    this.dayTimeEnds   = 20;
};

Randomizer.prototype = {
    every : function (x) {
        this.x = x;
        return this;
    },
    outOf : function (y) {
        this.y = y;
        return this;
    },
    get times () {
        return this;
    },
    defineDayTime : function (params) {
        this.dayTimeStarts = params.from;
        this.dayTimeEnds   = params.to;
    },
    getCurrentHour : function () {
        return this.moment().hours();
    },
    get ifItsDayTime () {
        var currentHour = this.getCurrentHour();
        this.isExecute = (currentHour >= this.dayTimeStarts && currentHour < this.dayTimeEnds);
        return this;
    },
    get ifItsNightTime () {
        var currentHour = this.getCurrentHour();
        this.isExecute = (currentHour < this.dayTimeStarts || currentHour >= this.dayTimeEnds);
        return this;
    },
    reset : function () {
        this.Math      = Math;
        this.moment    = moment;
        this.x         = null;
        this.y         = null;
        this.isExecute = true;
    },
    execute : function (fn) {
        // if only one of the probability values are set (or zero), throw an exception
        if ((this.x && !this.y) || (!this.x && this.y)) {
            this.reset();
            throw new Error('probability not set correctly');
        }
        var isChanceExecute = true;
        if (this.x && this.y) {
            var chance = (this.x / this.y);
            isChanceExecute = (chance >= this.Math.random());
        }
        if (isChanceExecute && this.isExecute) {
            fn && fn();
        }
        this.reset();
        return this;
    },
};

module.exports = exports = new Randomizer;
